tree-sitter-stack-graphs/src/cli/util.rs

```diff
@@ -240,103 +241,209 @@ pub fn duration_from_seconds_str(s: &str) -> Result<Duration, anyhow::Error> {
     Ok(Duration::new(s.parse()?, 0))
 }
 
-pub struct FileStatusLogger<'a> {
+pub fn iter_files_and_directories<'a, P, IP>(
+    paths: IP,
+) -> impl Iterator<Item = (PathBuf, PathBuf, bool)> + 'a
+where
+    P: AsRef<Path> + 'a,
+    IP: IntoIterator<Item = P> + 'a,
+{
+    paths
+        .into_iter()
+        .filter_map(
+            |source_path| -> Option<Box<dyn Iterator<Item = (PathBuf, PathBuf, bool)>>> {
```

Comment:
The typing here along with the generic parameters combined with lifetimes above looks like it was fun to sort out! It's nice that these generic parameters can all share the same lifetime 😅

Reply:
That was a tricky one to figure out. To be honest, for the more complex lifetime situations, my process of figuring it out is often still throwing stuff at the compiler and seeing what sticks 😬. There's some more interesting things coming in one of the follow-up PRs :)